### ISSUE TITLE
Limit on failure only

### ISSUE THREAD

**Issue by micw** [alisaifee/flask-limiter#234]
Hello,
this is not a bug but a feature request or (if already implemented) a request for advice ;-)
I have the following use case which I'd like to implement. The limiter is used to restrict authentication requests to prevent brute-force attacks. So the /auth endpoint is rate-limit. Unfortunately, in many use cases, valid authentication requests are hitting the limits too.
I'd like to achive the following behaviour:
- if a request comes in, it needs to be checked if there is already a limit hit. if so, the request it limited/rejected
- if there's no limit, the request is executed
- if it succeeds, it's not count into the limit. if it fails, it must count for the limit.

(How) would this be possible to implement?

Best regards,
Michael.

**Comment by alisaifee** [comment]
Hi @micw. Sorry for the delay in getting back to you.
At the moment there is no straightforward way to achieve this out of the box in flask-limiter. You could accomplish what you're looking for by directly using the [limits](http://github.com/alisaifee/limits) library.

It feels like this could be a common use case though and I can try to give it a shot if you're still in need of the feature.

**Comment by micw** [comment]
Hello, thank you the answer. In mailu, the limiter is used exactly at one single method:
https://github.com/Mailu/Mailu/blob/master/core/admin/mailu/internal/views/auth.py
This is the central authentication endpoint.

How can I fix this by using the library? Is it just calling moving_window.test() at method entry and on failure calling moving_window.hit() ?

**Comment by micw** [comment]
@alisaifee I'd greatly appreciate if you could have a look at my implementation at https://github.com/Mailu/Mailu/pull/1278/files if I'm using the library the right way.

I also had the idea to reset the counter on successful login but that seems not be supported by the lib. What do you think about it?

**Comment by alisaifee** [comment]
@micw. Your implementation looks good. I've exposed the method to reset individual limits [here](https://limits.readthedocs.io/en/stable/api.html#limits.storage.Storage.clear). If you ensure your version for the `limits` package is `>1.4.1` you should be able to use it as: `self.limiter.clear(self.rate,"client-ip",clientip)` in your code.

After looking through your implementation I'd like to include this functionality into the extension itself. My initial idea is to expose two new argument when declaring a limit using the decorator(s):
 - `penalty_condition`: A function or lambda that takes the current `response` object as a parameter and if it returns `True` the rate limit will be hit. This function will be called during  `after_request` so for example based on your code you would implement this as: 
    ```python
    def penalty_callback(response):
        return ("Auth-Status" not in response.headers) or (response.headers["Auth-Status"]!="OK")
    ```
 - `penalty_exceptions`: A list of exception classes which when raised by a request will trigger the rate limit to be hit. As per your example, you would then be able to do something like creating an `AuthenticationFailed` subclass of `HTTPException` and raise that in your view when the headers don't contain the appropriate values, and then declare `penalty_exceptions=[AuthenticationFailed]`

Based on your pull request those could be used as:
```
@utils.limiter.limit(
    lambda: app.config["AUTH_RATELIMIT"],
    lambda: flask.request.headers["Client-Ip"],
    penalty_condition=lambda response: ("Auth-Status" not in response.headers) or (response.headers["Auth-Status"]!="OK")
)
def nginx_authentication():
   ...
```

Clearing the limit upon a certain condition would also be a callback provided as another argument to the limiter, for example:
```
@utils.limiter.limit(
    ...
    reset_condition=lambda response: 'Auth-Status' in response.headers and response.headers["Auth-Status"] == "OK"
)
```

The names of the arguments will probably change to something else but that's the general idea. If that works for you I'll try and get that done in a few days. If your pull request is time sensitive I could propose that you go with what you have and use the new `clear` method to reset the limit if needed. Once I'm done with the new functionality you could change to using `Flask-Limiter` again and nothing *should* change (💥)

**Comment by micw** [comment]
@alisaifee That's great news. I'm not sure if we switch back to the annotations or stay with the lib because another dev would like to use the limit somewhere within a different service too. I'll check that.

I also got feedback from others about resetting the limit. An attacker who knows a single login could use this to probe other log-ins: probe till limit minus one, successful login to reset the limit, probe again. So it would effectively workaround the limit in some scenarios. That should at least be mentioned in the description of the new function.

Best regards,
Michael.

**Comment by CedricCabessa** [comment]
Hey @alisaifee , I'd love to see `penalty_condition` being implemented ! :heart: 

Are you still planning to do it ? If you don't have time, I can consider drafting a PR with your help :-)

In my usecase, I must rate limit differently 200 response and 4xx

**Comment by alisaifee** [comment]
I've pushed an initial implementation to address this in #248. @CedricCabessa please take a look.
@micw not sure if you're still interested considering its been months :sweat: but .. just in case.

**Comment by alisaifee** [comment]
The feature is implemented by #248.
